vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+update(D) <- D + n_ID
+
+n_IR <- rbinom(I, p_IR * dt * (1 - p_death))
+n_ID <- rbinom(I, p_IR * dt * p_death)
+n_SI <- rbinom(S, p_SI * dt)
+
+initial(D) <- 0
+p_death <- user(0.05)
+```
+
+```{r}
+gen_death <- odin_dust("deaths.txt")
+```
+
+We can simulate some data to fit to, as before. We will also make the case data patchy, making it missing on some days:
+
+```{r layout="l-body-outset", paged.print=TRUE }
+# Set up model
+dt <- 0.25
+inv_dt <- 1/dt
+sir_death <- gen_death$new(data = list(dt = dt, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
+                   step = 0,
+                   n_particles = 1L,
+                   n_threads = 1L,
+                   seed = 2L)
+
+# Run model forward 100 days, using 4 steps a day
+# Calculate the number of new cases by counting how many susceptibles have been depleted between updates
+n_steps <- 100
+incidence <- rep(NA, length.out = n_steps)
+deaths <- rep(NA, length.out = n_steps)
+true_history_deaths <- array(NA_real_, c(sir_death$info()$len, 1, n_steps + 1))
+true_history_deaths[, 1, 1] <- sir_death$state()
+for (t in seq_len(n_steps)) {
+    state_start <- sir_death$state()
+    state_end <- sir_death$run(t * inv_dt)
+    true_history_deaths[, 1, t + 1] <- state_end
+    incidence[t] <- state_start[2, 1] - state_end[2, 1]
+    deaths[t] <- state_end[5, 1] - state_start[5, 1]
+}
+
+# Add very simple form of noise
+missing <- rbinom(n = n_steps, prob = 0.05, size = 1) == 1
+incidence[missing] <- NA
+combined_data <- particle_filter_data(
+  data.frame(cases = incidence, deaths = deaths, day = seq_len(n_steps)),
+  time = "day",
+  rate = inv_dt
+)
+paged_table(combined_data)
+```
+
+We then need to add this into our comparison function, and in doing so we will also demonstrate the use of the `index` parameter for the particle filter. The only model outputs needed from the comparison function are $S$ and $D$. For models with a very large state space returning only those compartments required by the comparison function can be important for efficiency. An index function can be defined, which operates on dust generators `info()`:
+
+```{r}
+index <- function(info) {
+  list(run = c(info$index$S, info$index$D), state = c(info$index$S, info$index$I, info$index$R, info$index$D))
+}
+
+index(sir_death$info())
+```
+
+Which sends the comparison function the indices listed under `run`, in this case the number of susceptibles, and saves any compartments listed under `state` if the history is kept. In the `compare()` function the index of `state` now corresponds to `index$run`. The missing data must be checked for, and contribute 0 to the likelihood if found:
+
+```{r}
+combined_compare <- function(state, prev_state, observed, pars = NULL) {
+  if (is.na(observed$cases) && is.na(observed$deaths)) {
+    return(NULL)
+  }
+  if (is.null(pars$compare$exp_noise)) {
+    exp_noise <- 1e6
+  } else {
+    exp_noise <- pars$compare$exp_noise
+  }
+  
+  ll_pois <- function(obs, model, exp_noise) {
```

Comment:
This function might be lifted out to be a free function for ease of explanation - looks like it does not need the captured scope?